atomic_reactor/plugins/pre_pin_operator_digest.py

```diff
@@ -50,25 +53,187 @@ def __init__(self, tasker, workflow, replacement_pullspecs=None):
+            pullspecs.update(operator_csv.get_pullspecs())
+
+        if pullspecs:
+            pullspec_lines = "\n".join(sorted(map(str, pullspecs)))
+            self.log.info("Found pullspecs:\n%s", pullspec_lines)
+        else:
+            self.log.info("No pullspecs found")
+
+        return pullspecs
+
+    def _get_replacement_pullspecs(self, pullspecs):
+        self.log.info("Computing replacement pullspecs")
+        replacer = PullspecReplacer(user_config=self.user_config, site_config=self.site_config)
+
+        for p in pullspecs:
+            if not replacer.registry_is_allowed(p):
+                raise RuntimeError("Registry not allowed: {} (in {})".format(p.registry, p))
+
+        replacements = {}
+
+        for original in pullspecs:
+            pinned = replacer.pin_digest(original)
+            replaced = replacer.replace_registry(pinned)
+            if replaced != original:
+                replacements[original] = replaced
+
+        replacement_lines = "\n".join(
+            "{} -> {}".format(p, replacements[p]) if p in replacements
+            else "{} - no change".format(p)
+            for p in sorted(pullspecs, key=str)  # ImageName does not implement ordering
+        )
+        self.log.info("To be replaced:\n%s", replacement_lines)
+
+        return replacements
+
+    def _set_worker_arg(self, replacement_pullspecs):
+        arg = {str(old): str(new) for old, new in replacement_pullspecs.items()}
+        override_build_kwarg(self.workflow, "operator_bundle_replacement_pullspecs", arg)
+
+
+_KEEP = object()
+
+
+class PullspecReplacer(object):
+    """
+    Helper that takes care of replacing parts of image pullspecs
+    """
+
+    def __init__(self, user_config, site_config):
+        """
+        Initialize a PullspecReplacer
+
+        :param user_config: container.yaml operator_manifest configuration
+        :param site_config: reactor-config-map operator_manifests configuration
+        """
+        self.allowed_registries = site_config["allowed_registries"]
+        if self.allowed_registries is not None:
+            self.allowed_registries = set(self.allowed_registries)
+
+        # TODO: what if a replacement registry is not in allowed_registries?
+        self.registry_replace = {
+            registry["old"]: registry["new"]
+            for registry in site_config.get("registry_post_replace", [])
+        }
+
+    def registry_is_allowed(self, image):
+        """
+        Is image registry allowed in OSBS config?
+
+        :param image: ImageName
+        :return: bool
+        """
+        return self.allowed_registries is None or image.registry in self.allowed_registries
+
+    def pin_digest(self, image):
+        """
+        Replace image tag with manifest list digest
+
+        :param image: ImageName
+        :return: ImageName
+        """
+        if image.tag.startswith("sha256:"):
+            return image
+        # TODO: authentication? insecure?
```

Comment:
Ignore this for one now, we will resolve auth together with support for mutliple registries